## My Final Project
By Spencer Murphy

**Created: 4/24/21**

**Last Updated: 4/28/21**


For my final project, I'm choosing to construct a Pokemon "dream team" based on a number of stats like attack, speed, defense, etc. The question I will be attempting to answer is "what does an ideal pokemon dream team look like?" I will also be taking into account the movesets and abilities of the Pokemon found in the file. The file containing the data I will be using was found through this link: https://www.kaggle.com/rounakbanik/pokemon. The data comes straight from Pokemon.com and contains about 800 pokemon ranging from Gen 1 to Gen 7.

In terms of my analysis, I first plan to create a data frame using pandas that will exlcude many of the uncessary keys that are found in my data set. After I create a more polished and straightforward table to work with, I plan to graph the stats for all pokemon included in my data set to see the highest ranked pokemon in each category. Luckily, within my data set, there is a base_total key that is essentially a representation of how powerful a certain pokemon is. The higher a base_total is, the stronger the pokemon is. I plan to narrow down the top 30 strongest pokemon based on their base totals. I also plan to examine the history of the strongest pokemon in the many books, tv shows, and games to see which ones are the most canonically strong. But, there is a problem of legendary pokemon, which are known to be stronger than regular. Naturally, one would think that a team of all legendaries would be the closest thing to a "dream team," but I think that's pretty boring. So, I plan to make 2 teams - one consisting of all legendaries and one consisting of regular pokemon just to be fair.

In [142]:
import pandas as pd
from matplotlib import pyplot as plt

In [79]:
poke = pd.read_csv('pokemon.csv')
poke.keys()

Index(['abilities', 'against_bug', 'against_dark', 'against_dragon',
       'against_electric', 'against_fairy', 'against_fight', 'against_fire',
       'against_flying', 'against_ghost', 'against_grass', 'against_ground',
       'against_ice', 'against_normal', 'against_poison', 'against_psychic',
       'against_rock', 'against_steel', 'against_water', 'attack',
       'base_egg_steps', 'base_happiness', 'base_total', 'capture_rate',
       'classfication', 'defense', 'experience_growth', 'height_m', 'hp',
       'japanese_name', 'name', 'percentage_male', 'pokedex_number',
       'sp_attack', 'sp_defense', 'speed', 'type1', 'type2', 'weight_kg',
       'generation', 'is_legendary'],
      dtype='object')

Above are the many keys that are included in my data set. Like I said at the beginning, some of these keys will be eliminated as they won't mean much to my final results. For example, I will probably choose to exclude keys like pokedex_number, percentage_male, and generation.Considering my data is a csv file, we can see my data set as a table below. 

In [60]:
poke.head(10)



abilities  against_bug  against_dark  against_electric  \
0  ['Overgrow', 'Chlorophyll']         1.00           1.0               0.5   
1  ['Overgrow', 'Chlorophyll']         1.00           1.0               0.5   
2  ['Overgrow', 'Chlorophyll']         1.00           1.0               0.5   
3     ['Blaze', 'Solar Power']         0.50           1.0               1.0   
4     ['Blaze', 'Solar Power']         0.50           1.0               1.0   
5     ['Blaze', 'Solar Power']         0.25           1.0               2.0   
6     ['Torrent', 'Rain Dish']         1.00           1.0               2.0   
7     ['Torrent', 'Rain Dish']         1.00           1.0               2.0   
8     ['Torrent', 'Rain Dish']         1.00           1.0               2.0   
9  ['Shield Dust', 'Run Away']         1.00           1.0               1.0   

   against_fairy  against_fight  against_fire  against_flying  against_ghost  \
0            0.5            0.5           2.0             2.0            1.0   
1            0.5            0.5           2.0             2.0            1.0   
2            0.5            0.5           2.0             2.0            1.0   
3            0.5            1.0           0.5             1.0            1.0   
4            0.5            1.0           0.5             1.0            1.0   
5            0.5            0.5           0.5             1.0            1.0   
6            1.0            1.0           0.5             1.0            1.0   
7            1.0            1.0           0.5             1.0            1.0   
8            1.0            1.0           0.5             1.0            1.0   
9            1.0            0.5           2.0             2.0            1.0   

   against_grass  ...  percentage_male  pokedex_number  sp_attack  sp_defense  \
0           0.25  ...             88.1               1         65          65   
1           0.25  ...             88.1               2         80          80   
2           0.25  ...             88.1               3        122         120   
3           0.50  ...             88.1               4         60          50   
4           0.50  ...             88.1               5         80          65   
5           0.25  ...             88.1               6        159         115   
6           2.00  ...             88.1               7         50          64   
7           2.00  ...             88.1               8         65          80   
8           2.00  ...             88.1               9        135         115   
9           0.50  ...             50.0              10         20          20   

   speed  type1   type2  weight_kg  generation  is_legendary  
0     45  grass  poison        6.9           1             0  
1     60  grass  poison       13.0           1             0  
2     80  grass  poison      100.0           1             0  
3     65   fire     NaN        8.5           1             0  
4     80   fire     NaN       19.0           1             0  
5    100   fire  flying       90.5           1             0  
6     43  water     NaN        9.0           1             0  
7     58  water     NaN       22.5           1             0  
8     78  water     NaN       85.5           1             0  
9     45    bug     NaN        2.9           1             0  

[10 rows x 39 columns]

Here is an edited data frame that has the keys that I think are useful in reaching my conclusion. 

In [126]:
pok = poke.drop(labels = ['against_bug', 'against_dark', 'against_dragon', 'against_electric', 'against_fairy', 'against_fight', 'against_fire', 'against_flying', 'against_flying', 'against_ghost', 'against_grass', 'against_ground', 'against_ice', 'against_normal', 'against_poison', 'against_psychic', 'against_rock', 'against_steel', 'against_water', 'base_egg_steps', 'base_happiness', 'capture_rate', 'experience_growth', 'japanese_name', 'percentage_male', 'pokedex_number', 'type1', 'type2', 'generation'], axis = 1)
pok.head(10)

abilities  attack  base_total        classfication  \
0  ['Overgrow', 'Chlorophyll']      49         318         Seed Pokémon   
1  ['Overgrow', 'Chlorophyll']      62         405         Seed Pokémon   
2  ['Overgrow', 'Chlorophyll']     100         625         Seed Pokémon   
3     ['Blaze', 'Solar Power']      52         309       Lizard Pokémon   
4     ['Blaze', 'Solar Power']      64         405        Flame Pokémon   
5     ['Blaze', 'Solar Power']     104         634        Flame Pokémon   
6     ['Torrent', 'Rain Dish']      48         314  Tiny Turtle Pokémon   
7     ['Torrent', 'Rain Dish']      63         405       Turtle Pokémon   
8     ['Torrent', 'Rain Dish']     103         630    Shellfish Pokémon   
9  ['Shield Dust', 'Run Away']      30         195         Worm Pokémon   

   defense  height_m  hp        name  sp_attack  sp_defense  speed  weight_kg  \
0       49       0.7  45   Bulbasaur         65          65     45        6.9   
1       63       1.0  60     Ivysaur         80          80     60       13.0   
2      123       2.0  80    Venusaur        122         120     80      100.0   
3       43       0.6  39  Charmander         60          50     65        8.5   
4       58       1.1  58  Charmeleon         80          65     80       19.0   
5       78       1.7  78   Charizard        159         115    100       90.5   
6       65       0.5  44    Squirtle         50          64     43        9.0   
7       80       1.0  59   Wartortle         65          80     58       22.5   
8      120       1.6  79   Blastoise        135         115     78       85.5   
9       35       0.3  45    Caterpie         20          20     45        2.9   

   is_legendary  
0             0  
1             0  
2             0  
3             0  
4             0  
5             0  
6             0  
7             0  
8             0  
9             0

Below is proof that the base total is essentially a determiner for a pokemon's overall power. Pokemon evolve, and when they evolve they get stronger. Instance 0 is Bulbasaur, a sort of turtle pokemon that has a base total of 318. Instance 1 is Bulbasaur's next evolution, Ivysaur. It clearly has a higher base total at 405. Finally, Instance 2 is Bulbasaur's second and final evolution, Venusaur, who has a base total of 625. This pattern can be seen with every pokemon who harbors the ability to evolve.


In [127]:
poke['base_total'].head()

0    318
1    405
2    625
3    309
4    405
Name: base_total, dtype: int64

Here I've elminated all of the keys besides the names of the pokemon, their base totals, and legendary status (1 is yes, 0 is no) 

In [158]:
pokefix = pok.drop(labels = ['height_m', 'abilities', 'attack', 'classfication', 'hp', 'defense', 'sp_attack', 'sp_defense', 'weight_kg', 'speed'], axis = 1)
pokefix.head(10)

base_total        name  is_legendary
0         318   Bulbasaur             0
1         405     Ivysaur             0
2         625    Venusaur             0
3         309  Charmander             0
4         405  Charmeleon             0
5         634   Charizard             0
6         314    Squirtle             0
7         405   Wartortle             0
8         630   Blastoise             0
9         195    Caterpie             0

Below we can see the top 50 strongest pokemon according to their base totals. 

In [159]:
pokefix.sort_values('base_total', ascending = False)[:50]

base_total        name  is_legendary
149         780      Mewtwo             1
383         780    Rayquaza             1
382         770     Groudon             1
381         770      Kyogre             1
492         720      Arceus             1
717         708     Zygarde             1
380         700      Latios             1
372         700   Salamence             0
247         700   Tyranitar             0
444         700    Garchomp             0
718         700     Diancie             1
379         700      Latias             1
375         700   Metagross             0
645         700      Kyurem             1
719         680       Hoopa             1
716         680     Yveltal             1
715         680     Xerneas             1
248         680       Lugia             1
643         680      Zekrom             1
791         680      Lunala             1
790         680    Solgaleo             1
642         680    Reshiram             1
249         680       Ho-Oh             1
482         680      Dialga             1
486         680    Giratina             1
483         680      Palkia             1
288         670     Slaking             0
485         670   Regigigas             1
657         640    Greninja             0
129         640    Gyarados             0
259         635    Swampert             0
5           634   Charizard             0
256         630    Blaziken             0
253         630    Sceptile             0
8           630   Blastoise             0
305         630      Aggron             0
447         625     Lucario             0
2           625    Venusaur             0
745         620  Wishiwashi             0
474         618     Gallade             0
281         618   Gardevoir             0
141         615  Aerodactyl             0
207         610     Steelix             0
180         610    Ampharos             0
484         600     Heatran             1
64          600    Alakazam             0
385         600      Deoxys             1
250         600      Celebi             1
93          600      Gengar             0
720         600   Volcanion             1